gcp-pubsub/src/main/java/io/micronaut/gcp/pubsub/intercept/PubSubClientIntroductionAdvice.java

```diff
@@ -0,0 +1,173 @@
+import io.reactivex.Scheduler;
+import io.reactivex.Single;
+import io.reactivex.schedulers.Schedulers;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Named;
+import javax.inject.Singleton;
+import java.util.*;
+import java.util.concurrent.ExecutorService;
+
+/**
+ * Implementation of {@link io.micronaut.gcp.pubsub.annotation.PubSubClient} advice annotation.
+ *
+ * @author Vinicius Carvalho
+ * @since 2.0.0
+ */
+@Singleton
+public class PubSubClientIntroductionAdvice implements MethodInterceptor<Object, Object> {
+
+    private final Logger logger = LoggerFactory.getLogger(PubSubClientIntroductionAdvice.class);
+    private final PublisherFactory publisherFactory;
+    private final PubSubMessageSerDesRegistry serDesRegistry;
+    private final Scheduler scheduler;
+    private final ConversionService<?> conversionService;
+
+    public PubSubClientIntroductionAdvice(PublisherFactory publisherFactory,
+                                          PubSubMessageSerDesRegistry serDesRegistry,
+                                          @Named(TaskExecutors.IO) ExecutorService executorService,
+                                          ConversionService<?> conversionService) {
+        this.publisherFactory = publisherFactory;
+        this.serDesRegistry = serDesRegistry;
+        this.scheduler = Schedulers.from(executorService);
+        this.conversionService = conversionService;
+    }
+
+    @Override
+    public Object intercept(MethodInvocationContext<Object, Object> context) {
+
+        if (context.hasAnnotation(Topic.class)) {
+            Map<String, String> messageAttributes = new HashMap<>();
+            AnnotationValue<Topic> topic = context.getAnnotation(Topic.class);
+            String contentType = topic.get("contentType", String.class).orElse("");
+            List<AnnotationValue<Header>> headerAnnotations = context.getAnnotationValuesByType(Header.class);
+
+            PubSubMessageSerDes serDes = serDesRegistry.find(contentType)
+                    .orElseThrow(() -> new PubSubClientException("Could not locate a valid SerDes implementation for type: " + contentType));
+            Publisher publisher = publisherFactory.createPublisher(topic.getValue(String.class).orElse(""));
+            Argument<?> bodyArgument = findBodyArgument(context.getExecutableMethod())
+                    .orElseThrow(() -> new PubSubClientException("No valid message body argument found for method: " + context.getExecutableMethod()));
+
+            Map<String, Object> parameterValues = context.getParameterValueMap();
+
+            headerAnnotations.forEach((header) -> {
+                String name = header.get("name", String.class).orElse(null);
+                String value = header.getValue(String.class).orElse(null);
+                if (StringUtils.isNotEmpty(name) && StringUtils.isNotEmpty(value)) {
+                    messageAttributes.put(name, value);
+                }
+            });
```

Comment:
Some of these things can be cached. The header annotations, body argument, topic/serdes will not change between executions.

Reply:
Its on my todo list, will change for this PR